### PR TITLE
Feature/446

### PR DESCRIPTION
**Pull request by sebabarre**
Distinct authentication and token/jwk handling.
Token and jwk are now generated without waiting for an incoming request from user

### CODE REVIEW

File: bot/connector-teams/src/main/kotlin/token/TokenHandler.kt [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+object TokenHandler {
+
+    private val logger = KotlinLogging.logger {}
+
+    @Volatile
+    var token: String? = null
+    @Volatile
+    private var tokenExpiration: Instant? = null
+
+    private lateinit var appId: String
```
Comment (vsct-jburet) [inline bot/connector-teams/src/main/kotlin/token/TokenHandler.kt]: appId and password should be Volatile too

File: bot/connector-teams/src/main/kotlin/token/TokenHandler.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    private lateinit var password: String
+
+    val teamsMapper: ObjectMapper = mapper.copy().setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE)
+
+
+    var loginApi: LoginMicrosoftOnline = retrofitBuilderWithTimeoutAndLogger(
+        longProperty("tock_whatsapp_request_timeout_ms", 30000),
```
Comment (vsct-jburet) [inline bot/connector-teams/src/main/kotlin/token/TokenHandler.kt]: tock_whatsapp_.. is not a good property name ;)

File: bot/connector-teams/src/main/kotlin/TeamsConnector.kt [hunk 3]
```diff
@@ -47,70 +50,76 @@ internal class TeamsConnector(
...
         private val logger = KotlinLogging.logger {}
     }
 
-    private val client = TeamsClient(appId, appPassword)
+    private val client = TeamsClient()
     private val executor: Executor by injector.instance()
     private val authenticateBotConnectorService = AuthenticateBotConnectorService(appId)
 
     override fun register(controller: ConnectorController) {
+        launchTokenCollector(appId, appPassword)
```
Comment (vsct-jburet) [inline bot/connector-teams/src/main/kotlin/TeamsConnector.kt]: you should stop the crons in a Connector#unregister method

File: bot/connector-teams/src/main/kotlin/token/TokenHandler.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    fun setId(appId: String, password: String) {
+        this.appId = appId
+        this.password = password
+    }
+
+    fun launchTokenCollector(appId: String, password: String, msInterval: Long = (60 * 60 * 1000).toLong()) {
+        setId(appId, password)
+        fixedRateTimer(name = "microsoft-api-token-handling", initialDelay = 0.toLong(), period = msInterval) {
```
Comment (vsct-jburet) [inline bot/connector-teams/src/main/kotlin/token/TokenHandler.kt]: initialDelay = 0L is better

File: bot/connector-teams/src/main/kotlin/token/TokenHandler.kt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        token = response.body()?.accessToken ?: error("empty access token")
+        tokenExpiration = Instant.now().plus(response.body()?.expiresIn!!, ChronoUnit.SECONDS)
+    }
+
+    fun setId(appId: String, password: String) {
+        this.appId = appId
+        this.password = password
+    }
+
+    fun launchTokenCollector(appId: String, password: String, msInterval: Long = (60 * 60 * 1000).toLong()) {
```
Comment (vsct-jburet) [inline bot/connector-teams/src/main/kotlin/token/TokenHandler.kt]: msInterval: Long = 60 * 60 * 1000L is better

File: bot/connector-teams/src/main/kotlin/auth/JWKHandler.kt [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            collectJWK()
+        }
+    }
+
+    private fun collectJWK() {
+        logger.debug("Getting new jwks")
+        val microsoftOpenidMetadata = microsoftOpenIdMetadataApi.getMicrosoftOpenIdMetadata().execute()
+        val response = microsoftOpenidMetadata.body()
+        tokenIds = response?.idTokenSigningAlgValuesSupported
+            ?: throw IOException("Error : Unable to get OpenidMetadata to validate BotConnectorServiceKeys")
```
Comment (vsct-jburet) [inline bot/connector-teams/src/main/kotlin/auth/JWKHandler.kt]: I prefer to use error("error message") that throws an IllegalStateException

File: bot/connector-teams/src/main/kotlin/token/TokenHandler.kt (hunk 1 above)
Comment (sebabarre) [reply bot/connector-teams/src/main/kotlin/token/TokenHandler.kt]: good point

File: bot/connector-teams/src/main/kotlin/TeamsConnector.kt (hunk 3 above)
Comment (sebabarre) [reply bot/connector-teams/src/main/kotlin/TeamsConnector.kt]: Very good point

**Comment by vsct-jburet** [comment]
merged

**Pull request closed without merging**